File: 945.minimum-increment-to-make-array-unique.js
/*
 * @lc app=leetcode id=945 lang=javascript
 *
 * [945] Minimum Increment to Make Array Unique
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var minIncrementForUnique = function (nums)
{
    nums.sort((a, b) => a - b);

    let step = 0;

    for (let i = 1; i < nums.length; i++)
    {
        if (nums[i] === nums[i - 1])
        {
            let num = nums[i];
            let j = i;
            while (j < nums.length && nums[j] <= num)
            {
                nums[j] = nums[j + 1];
                num++;
                j++;
            }
            nums[j - 1] = num;
            step += j - i;
            i--;
        }
    }

    return step;
};
// @lc code=end

File: Daily/945.minimum-increment-to-make-array-unique.cpp
/*
 * @lc app=leetcode id=945 lang=cpp
 *
 * [945] Minimum Increment to Make Array Unique
 */
#include <algorithm>
#include <vector>

// @lc code=start
class Solution {
 public:
  int minIncrementForUnique(std::vector<int>& nums) {
    std::sort(nums.begin(), nums.end());
    int lastUniqueNum = -1;
    int incrementNumber = 0;
    for (const int num : nums) {
      if (num > lastUniqueNum) {
        lastUniqueNum = num;
      } else {
        lastUniqueNum++;
        incrementNumber += lastUniqueNum - num;
      }
    }

    return incrementNumber;
  }
};
// @lc code=end
